ass/document.py

```diff
@@ -578,3 +432,247 @@ class Command(_Event):
     """ A command event. Not widely supported.
     """
     TYPE = "Command"
+
+class CaseInsensitiveOrderedDict(collections.abc.MutableMapping):
+    """A case insensitive ordered dictionary that preserves the original casing."""
+
+    def __init__(self, *args, **kwargs):
+        self._dict = collections.OrderedDict(*args, **kwargs)
```

Comment:
You could have also just subclassed `OrderedDict` and overridden `__*item__` and `__contains__` with `return super().__*__(key.lower(), *args)`

Reply:
True, I can do that instead. Would save us from having to implement `__iter__`, `__len__`, `__repr__` and `__str__`, although we would probably have to implement `__contains__` explicitly instead.

I'd rather keep the current general design though with the auxiliary `_case_mapping` as opposed to just making all keys lower case, as I think it's nice to get the original casing when iterating over keys.